Add async queue processing

thenbrent opened the pull request:
This is part two of #319 - the async queue processing. Part one - async action APIs - can be found in #322.

This PR extends the default `ActionScheduler_QueueRunner` class to include async processing. Queues are now run in asynchronous requests, as well as WP Cron requests.

These requests are initiated via the 'shutdown' hook on admin requests when Action Scheduler is not already processing the maximum number of allowed queues. The `'shutdown'` hook is the same approach taken by Jetpack and [Async Task](https://github.com/techcrunch/wp-async-task).

This removes Action Scheduler's dependence on WP Cron. While `ActionScheduler_QueueRunner` still uses WP Cron to process queues, queues will also now be processed anytime an admin request is made and there are pending actions.

Beyond removing the dependency on WP Cron, this has another major advantage - speed.

The approach taken here includes async request chaining (as initially discussed in #131). Whenever one async request processes a batch of actions to the allowed limits (the 20 second time limit is normally the limit that would be hit), a new async request to start processing a new batch of actions will be initiated if there are still other async actions to process.

This reduces the lag between processing each batch, and combined with queue concurrency, it substantially increases throughput when more than one queue is being processed by async requests (especially given the fact that each async request is also an admin request, so it will also initiate a 2nd async request on `'shutdown'` in addition to the async request created to chain processing for the original request).

Importantly, this differs substantially to the process currently used with WP Cron, which is to simply process as many actions/batches as possible within a given time limit during that single request, then wait for another WP Cron request before starting another batch.

### Implementation Notes

#### Keeping WP Cron

I've kept the WP Cron processing, because it's a helpful fallback for when loopback requests are not supported by the site/server.

#### Running Queues via the `'action_scheduler_run_queue'` Action

Importantly, queues are run in the async request by triggering the ` 'action_scheduler_run_queue'` hook, rather than calling `ActionScheduler::runner()->run()` directly.

This maintains compatibility with the WP Cron runner, and helps ensure existing code will apply to this new method for processing queues, like this plugin: https://github.com/Prospress/action-scheduler-disable-default-runner

#### Reduced Default Queue Concurrency

Importantly, given the increased likelihood of concurrent queues now occurring on all sites (instead of just sites tweaking things with Action Scheduler), there is a need to be more conservative with the default concurrency allowed by Action Scheduler.

SHA: f493e7d9914b5ad7e66d5 takes care of this, by reducing the default concurrency to two queues, from five. This prevents consuming too many resources by default. The concurrency can still be [increased](https://actionscheduler.org/perf/#increasing-concurrent-batches) when possible, if needed. I expect the chained async requests will increase throughput enough that concurrency increases will be required less now.

#### Disabling Async Queues with `'action_scheduler_allow_async_runner'`

As mentioned [here](https://github.com/Prospress/action-scheduler/issues/131#issuecomment-391227809), chained async requests without a kill-switch can be a dangerous thing.

To mitigate this, I've included a new `'action_scheduler_allow_async_runner'` filter. This can be used to both:

1. prevent any actions being run via async requests
1. terminate an async request chain part way through

- [ ] this new filter will need to be added to [the docs](https://actionscheduler.org/).

#### Initialising the `ActionScheduler_AsyncRequest_QueueRunner`

I'm injecting the `ActionScheduler_AsyncRequest_QueueRunner` class into the `ActionScheduler_QueueRunner`. That's to match the approach taken for other dependencies by `ActionScheduler_QueueRunner`; however, I don't see much need to swap it.

#### `WP_Async_Request`

I've used the `WP_Async_Request` library used by [WP Background Processing](https://github.com/deliciousbrains/wp-background-processing) for handling the request. This wasn't really necessary, as its mostly just a wrapper around `wp_remote_post()`, but I figured using it was better than starting from scratch.

## Testing

A fun way to test this is with the following snippet and visiting a page on your site with the URL param `?as_add_async_actions=true`. It will require the patch in #322 to get the `as_enqueue_async_action()` method.

```php
function as_test_schedule_async_action() {

	if ( ! isset( $_GET['as_add_async_actions'] ) ) {
		return;
	}

	for( $i = 1; $i < 300; $i++ ) {
		as_enqueue_async_action( 'test_async_action', array( 'timestamp' => time(), 'i' => $i ), 'async' );
	}
}
add_action( 'shutdown', 'as_test_schedule_async_action' );

function as_test_schedule_async_action_sleep() {
	sleep( 1 );
}
add_action( 'test_async_action', 'as_test_schedule_async_action_sleep' );
```

This will add hundreds of actions, and process them at a sustained rate of 120 / minute, as each action takes 1 second to process, and two concurrent queues will be processing them. With WP Cron only, the maximum throughput would be 20 / minute.

rrennick commented:
This PR needs #322 to be merged before being tested.

rrennick commented:
After merging in #322, my async actions aren't running on dashboard loads. I'm seeing one of these warnings for each dashboard load:

```
[05-Jul-2019 13:37:55 UTC] PHP Notice:  Undefined property: ActionScheduler_AsyncRequest_QueueRunner::$data in wp-content/plugins/action-scheduler/lib/WP_Async_Request.php on line 107
```

thenbrent commented:
>  I'm seeing one of these warnings for each dashboard load

Good catch. That was a potentially nasty bug. I had modified the `WP_Async_Request` class while checking how much of it was actually needed (and considering creating a more lightweight fork) and mistakenly committed those modifications under `WP_Async_Request`. That would lead to compatibility issues with any 3rd party using that class, like WC core and many others. I've updated the `WP_Async_Request` to be an exact copy of https://github.com/deliciousbrains/wp-background-processing/blob/fbbc56f2480910d7959972ec9ec0819a13c6150a/classes/wp-async-request.php

I've also changed the base branch of this PR to `version_3_0_0` and cherry-picked the commits to be on top of that branch, instead of `master` for a clean merge.

thenbrent commented:
I've also just added

* SHA: 2eb84ff54a049 to improve the location of the `'action_scheduler_allow_async_request_runner'` filter; and
* SHA: ee788f5df62 to avoid introducing an issue like https://github.com/woocommerce/woocommerce-admin/issues/2591

rrennick commented:
@thenbrent This worked great for me in testing. The one thing I noted was that if you load the Scheduled Actions admin screen it includes the pending actions that will be processed when the page load completes. Could we add a notice to the top of that screen indicating that some actions may be run asynchronously?

thenbrent commented:
Good idea @rrennick. I've added that with SHA: ec1ec246b1072. The locking code required quite a bit of refactoring to be able to provide effective notices.

Here is an example of the notice now displayed when there are pending actions due to be processed and:

* the lock is not set, so a new async runner will be triggered on `'shutdown'`: https://cld.wthms.co/uKKCgS
* the async queue runner lock is set: https://cld.wthms.co/olUH8t

This system isn't perfect, because the `in-progress` view will only show actions with that status at the time of the page load (and it's possible AS will be between actions during that time), but it's still a nice improvement.

rrennick commented on classes/ActionScheduler_OptionLock.php:
```diff
@@ -0,0 +1,49 @@
+	 * @bool True if lock value has changed, false if not or if set failed.
+	 */
+	public function set( $lock_type ) {
+		return update_option( $this->get_key( $lock_type ), time() + $this->get_duration( $lock_type ), $this->get_duration( $lock_type ) );
```
```suggestion
		return update_option( $this->get_key( $lock_type ), time() + $this->get_duration( $lock_type ) );
```

The third parameter of `update_option` is the `autoload` value.

rrennick commented on classes/abstracts/ActionScheduler_Lock.php:
```diff
@@ -0,0 +1,60 @@
+	 * @param string $lock_type A string to identify different lock types.
+	 * @return bool
+	 */
+	public function set( $lock_type ) {}
```
Should this be abstract?

rrennick commented on classes/abstracts/ActionScheduler_Lock.php:
```diff
@@ -0,0 +1,60 @@
+	 * @param string $lock_type A string to identify different lock types.
+	 * @return bool|int False if no lock is set, otherwise the timestamp for when the lock is set to expire.
+	 */
+	public function get_expiration( $lock_type ) {}
```
Also abstract?

rrennick commented:
@thenbrent Other than the 3 small things I noted, this works great. I checked running AS Admin and the REST API calls will process the pending actions in the background. I may want to add similar notices to it.

thenbrent commented:
@rrennick good catch on all 3 counts! I've rebased in fixes. :)

rrennick commented on classes/abstracts/ActionScheduler_Lock.php:
```diff
@@ -0,0 +1,60 @@
+<?php
+
+/**
+ * Class 
```
```suggestion
 * Class
```

Remove trailing space. Should we look at getting a code sniffer set up for this repo?

thenbrent replied:
> Should we look at getting a code sniffer set up for this repo?

That would be great. The question on that is which set of standards should it follow? It definitely doesn't follow WP standards yet.

JPry commented on classes/ActionScheduler_AsyncRequest_QueueRunner.php:
```diff
@@ -0,0 +1,74 @@
+ * ActionScheduler_AsyncRequest_QueueRunner
+ */
+
+defined( 'ABSPATH' ) || exit;
```
There's nothing wrong with including this, but I just wanted to note that as long as this file (or any other file) *only* declares a class with no other secondary logic, it's OK to omit this check.

thenbrent commented:
> Feel free to merge once the space is removed.

I rebased in a patch to add the class, and also a brief comment on the purpose of the class. I'll merge once Travis passes. 🚀
